src/nvim/quickfix.c

```diff
@@ -1154,35 +1241,259 @@ static char_u * qf_cmdtitle(char_u *cmd)
   return qftitle_str;
 }
 
+// Return a pointer to the current list in the specified quickfix stack
+static qf_list_T * qf_get_curlist(qf_info_T *qi)
+{
+  return qf_get_list(qi, qi->qf_curlist);
+}
+
 // Prepare for adding a new quickfix list. If the current list is in the
 // middle of the stack, then all the following lists are freed and then
 // the new list is added.
 static void qf_new_list(qf_info_T *qi, const char_u *qf_title)
 {
   int i;
+  qf_list_T *qfl;
 
   // If the current entry is not the last entry, delete entries beyond
   // the current entry.  This makes it possible to browse in a tree-like
   // way with ":grep'.
-  while (qi->qf_listcount > qi->qf_curlist + 1)
-    qf_free(qi, --qi->qf_listcount);
+  while (qi->qf_listcount > qi->qf_curlist + 1) {
+    qf_free(&qi->qf_lists[--qi->qf_listcount]);
+  }
 
   /*
    * When the stack is full, remove to oldest entry
    * Otherwise, add a new entry.
    */
   if (qi->qf_listcount == LISTCOUNT) {
-    qf_free(qi, 0);
-    for (i = 1; i < LISTCOUNT; ++i)
+    qf_free(&qi->qf_lists[0]);
+    for (i = 1; i < LISTCOUNT; i++) {
       qi->qf_lists[i - 1] = qi->qf_lists[i];
+    }
     qi->qf_curlist = LISTCOUNT - 1;
   } else
     qi->qf_curlist = qi->qf_listcount++;
-  memset(&qi->qf_lists[qi->qf_curlist], 0, (size_t)(sizeof(qf_list_T)));
-  qf_store_title(qi, qi->qf_curlist, qf_title);
-  qi->qf_lists[qi->qf_curlist].qf_id = ++last_qf_id;
+  qfl = qf_get_curlist(qi);
+  memset(qfl, 0, (size_t)(sizeof(qf_list_T)));
+  qf_store_title(qfl, qf_title);
+  qfl->qfl_type = qi->qfl_type;
+  qfl->qf_id = ++last_qf_id;
+}
+
+// Parse the match for filename ('%f') pattern in regmatch.
+// Return the matched value in "fields->namebuf".
```

Comment:
The comments above a function should be Doxygen-style comments, that is `///`. See http://doxygen.nl/manual/docblocks.html#cppblock